### PR TITLE
Added GCS bucket root folder into the output

### LINKED ISSUE #23: Add GCS bucket root folder into the output
We currently don't include GCS bucket root folder in the report which makes it hard to follow and identify sensitive buckets. We need to include them too.

### PR DESCRIPTION
**Pull request by ro4i7**
## Description
Fixes #23 

 It retrieves the root folder information for each bucket and adds it to a list within the `buckets_dict` value tuple.

## Changes Made
[List the changes you have made in bullet points.]

## Checklist
- [x] I have read and followed the contributing guidelines.
- [ ] I have tested my changes thoroughly and they work as expected.
- [ ] I have added necessary tests for the changes made.
- [ ] I have updated the documentation to reflect the changes made.
- [x] My code follows the project's coding style and standards.
- [x] I have added appropriate commit messages and comments for my changes.

## Related Issues
[If your changes relate to any existing issues, please list them here.]

## Additional Notes
[Provide any additional notes or context that may be helpful for the reviewers.]

Feel free to modify and customize this template as needed to fit the specific needs of the GCP Scanner project.

### CODE REVIEW

**Comment by ro4i7** [comment]
@mshudrak please check

File: src/gcp_scanner/crawl.py [hunk 1]
```diff
@@ -341,16 +341,15 @@ def get_bucket_names(project_name: str, credentials: Credentials,
                      dump_fd: io.TextIOWrapper
                      ) -> Dict[str, Tuple[Any, List[Any]]]:
   """Retrieve a list of buckets available in the project.
-
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:344]: Please keep blank line between sections

File: src/gcp_scanner/crawl.py [hunk 2]
```diff
@@ -341,16 +341,15 @@ def get_bucket_names(project_name: str, credentials: Credentials,
...
   Args:
     project_name: A name of a project to query info about.
     credentials: An google.oauth2.credentials.Credentials object.
     dump_fd: If set, the function will enumerate files stored in buckets and
       save them in a file corresponding to provided file descriptor.
       This is a very slow, noisy operation and should be used with caution.
-
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:351]: Please keep blank line between sections

File: src/gcp_scanner/crawl.py [hunk 3]
```diff
@@ -341,16 +341,15 @@ def get_bucket_names(project_name: str, credentials: Credentials,
...
   Returns:
-    A dictionary where key is bucket name and value is a bucket Object.
+    A dictionary where key is bucket name and value is a tuple containing
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:351]: "A dictionary with the bucket name as key containing a tuple of the bucket object and a list of root directory objects."

File: src/gcp_scanner/crawl.py [hunk 4]
```diff
@@ -368,23 +367,28 @@ def get_bucket_names(project_name: str, credentials: Credentials,
       break
 
     for bucket in response.get("items", []):
-      buckets_dict[bucket["name"]] = (bucket, None)
-      if dump_fd is not None:
-        ret_fields = "nextPageToken,items(name,size,contentType,timeCreated)"
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:373]: Why was this inside this if statement before and why is it outside now?

File: src/gcp_scanner/crawl.py [hunk 5]
```diff
@@ -368,23 +367,28 @@ def get_bucket_names(project_name: str, credentials: Credentials,
...
 
-            req = service.objects().list_next(req, resp)
-          except googleapiclient.errors.HttpError:
-            logging.info("Failed to read the bucket %s", bucket["name"])
-            logging.info(sys.exc_info())
-            break
+          req = service.objects().list_next(req, resp)
+        except googleapiclient.errors.HttpError:
+          logging.info("Failed to read the bucket %s", bucket["name"])
+          logging.info(sys.exc_info())
```
Comment (ZetaTwo) [inline src/gcp_scanner/crawl.py:390]: For log level info, dumping sys.exc_info seems a bit excessive

**Review by ZetaTwo — CHANGES_REQUESTED** [review]
Looks good overall. Just some minor comments

File: src/gcp_scanner/crawl.py (hunk 4 above)
Comment (ro4i7) [reply src/gcp_scanner/crawl.py:373]: in the original code, if a file descriptor is not provided, then the `ret_fields` and `req` variables are not needed, as the function does not need to retrieve and dump file information. But here in the modified code, the `ret_fields` and `req` variables are required to retrieve and store the root folder information for each bucket, regardless of whether a file descriptor has been provided. So this is moved outside here.

File: src/gcp_scanner/crawl.py (hunk 5 above)
Comment (ro4i7) [reply src/gcp_scanner/crawl.py:390]: Yeah, but this information can be useful for debugging purposes, in the future.

File: src/gcp_scanner/crawl.py (hunk 4 above)
Comment (mshudrak) [reply src/gcp_scanner/crawl.py:373]: Hm, I think there should be easier way to retrieve root folder of the given bucket. This way we generate much more traffic and noise

Comment (mshudrak) [reply src/gcp_scanner/crawl.py:373]: you already have it under the bucket['name']

Comment (mshudrak) [reply src/gcp_scanner/crawl.py:373]: The whole idea of dump_fd is to make user choose whether they want or not want to enable each file enumeration in the bucket to speed up the scanning. This change removes this ability. I suggest to keep it.

Comment (ro4i7) [reply src/gcp_scanner/crawl.py:373]: @mshudrak   I got it. I have modified the code according to it. Please check

File: src/gcp_scanner/crawl.py [hunk 6]
```diff
@@ -368,18 +369,29 @@ def get_bucket_names(project_name: str, credentials: Credentials,
       break
 
     for bucket in response.get("items", []):
-      buckets_dict[bucket["name"]] = (bucket, None)
+      buckets_dict[bucket["name"]] = (bucket, [])
+
+      if dump_fd is None:
```
Comment (mshudrak) [inline src/gcp_scanner/crawl.py:374]: Like I said in the previous comment. You already have everything including root folder name. All you need is to just add that data. No need for extra request. Take a look at what API returns you. Also could you please make sure to test your code before submitting PRs? I see extra null field in the response output.

Comment (ro4i7) [reply src/gcp_scanner/crawl.py:374]: Ok, working on it

**Review by mshudrak — CHANGES_REQUESTED** [review]
Please fix indentation first in order to proceed with changes.

File: src/gcp_scanner/crawl.py [hunk 7]
```diff
@@ -338,58 +338,67 @@ def get_firewall_rules(
...
-            resp = req.execute()
-            for item in resp.get("items", []):
-              dump_fd.write(json.dumps(item, indent=2, sort_keys=False))
-
-            req = service.objects().list_next(req, resp)
-          except googleapiclient.errors.HttpError:
-            logging.info("Failed to read the bucket %s", bucket["name"])
+                      dump_fd: Optional[io.TextIOWrapper] = None
+                      ) -> Dict[str, Tuple[Any, List[Any]]]:
+    """Retrieve a list of buckets available in the project.
```
[the remark this answers is not in the document]
Comment (ro4i7) [reply src/gcp_scanner/crawl.py:345]: Done

[the remark this answers is not in the document]
Comment (mshudrak) [reply src/gcp_scanner/crawl.py:345]: I don't think it is correct. PTAL at the original code.

**Comment by mshudrak** [comment]
Fixed in #150. Closing this one.
